// @ogod/core/dist/runtime/default/state/default-instance.js
import { OgodInstanceRuntime } from "../../../core/index.js";
export class DefaultInstanceRuntime extends OgodInstanceRuntime {
  addFeature(feature, entity) {
    this.features.push(feature);

    if (entity.initialized) {
      feature.initialize(this, entity);
    }
  }

  initialize(entity) {
    this.features.forEach(f => f.initialize(this, entity));
    return Object.assign({}, entity, {
      properties: this.changesPropertyAll(entity),
      initialized: true
    });
  }

  changes(entity) {
    this.features.forEach(f => entity = f.changes(this, entity));
    return Object.assign({}, entity);
  }

  changesPropertyAll(entity) {
    let properties = entity.properties;

    if (properties && Object.keys(properties).length > 0) {
      let newProps = {};
      Object.values(properties).forEach(state => {
        if (state.entity.binding) {
          let newState = this.changesProperty(entity, state);

          if (newState !== state) {
            newProps[state.entity.name] = newState;
          }
        }
      });

      if (Object.keys(newProps).length > 0) {
        return Object.assign({}, properties, newProps);
      }
    }

    return properties;
  }

  changesProperty(_entity, state) {
    let newEntity = state.runtime.update(state.entity);

    if (newEntity !== state.entity) {
      return Object.assign({}, state, {
        entity: newEntity
      });
    }

    return state;
  }

  update(entity, delta) {
    let e;
    this.features.forEach(f => {
      let fe = f.update(this, entity, delta);

      if (entity !== fe) {
        e = Object.assign({}, e, fe);
      }
    });
    return e || entity;
  }

} //# sourceMappingURL=default-instance.js.map
