// @ogod/core/dist/core/state/ogod-scene-changes.js
export class OgodSceneChanges {
  get additions() {
    return this._added;
  }

  get deletions() {
    return this._deleted;
  }

  get modifications() {
    return this._modified;
  }

  get hasAny() {
    return Object.keys(this._added).concat(Object.keys(this._modified)).concat(Object.keys(this._deleted)).length > 0;
  }

  constructor() {
    this._added = {};
    this._deleted = {};
    this._modified = {};
  }

  added(instance) {
    this._added[instance.id] = Object.assign({}, this._added[instance.id], instance);
    return this;
  }

  deleted(instance) {
    this._deleted[instance.id] = Object.assign({}, this._deleted[instance.id], instance);
    return this;
  }

  modified(id, instance) {
    this._modified[id] = Object.assign({}, this._modified[instance.id], instance);
    return this;
  }

  merge(changes) {
    this._added = Object.assign({}, this._added, changes._added);
    this._deleted = Object.assign({}, this._deleted, changes._deleted);
    this._modified = Object.assign({}, this._modified, changes._modified);
  }

} //# sourceMappingURL=ogod-scene-changes.js.map
