// @ogod/core/dist/core/state/ogod-scene-changes.js
import { mergeInstances } from "../utils.js";
export class OgodSceneChanges {
  constructor() {
    this._added = {};
    this._deleted = {};
    this._modified = {};
  }

  get additions() {
    return this._added;
  }

  get deletions() {
    return this._deleted;
  }

  get modifications() {
    return this._modified;
  }

  get hasAny() {
    return Object.keys(this._added).concat(Object.keys(this._modified)).concat(Object.keys(this._deleted)).length > 0;
  }

  added(instance) {
    this._added[instance.id] = Object.assign(Object.assign({}, this._added[instance.id]), instance);
    return this;
  }

  deleted(instance) {
    this._deleted[instance.id] = Object.assign(Object.assign({}, this._deleted[instance.id]), instance);
    return this;
  }

  modified(id, instance) {
    this._modified[id] = Object.assign(Object.assign({}, this._modified[instance.id]), instance);
    return this;
  }

  merge(changes) {
    this._added = mergeInstances(this._added, changes._added);
    this._deleted = mergeInstances(this._deleted, changes._deleted);
    this._modified = mergeInstances(this._modified, changes._modified);
  }

} //# sourceMappingURL=ogod-scene-changes.js.map
