// @ogod/runtime-three/dist/component/three-textures.js
var __decorate = this && this.__decorate || function (decorators, target, key, desc) {
  var c = arguments.length,
      r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
      d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
};

import { customElement, property } from "../../../../lit-element/lit-element.js";
import { OgodResourceElement } from "../../../core/dist/index.js";
import { ThreeTexturesRuntime } from "../state/three-textures-runtime.js";
let ThreeTexturesElement = class ThreeTexturesElement extends OgodResourceElement {
  constructor() {
    super();
    this.start = 0;
    this.end = 0;

    this.impl = () => new ThreeTexturesRuntime();
  }

  getEntity() {
    return Object.assign({}, super.getEntity(), {
      startIndex: this.start,
      endIndex: this.end,
      extension: this.ext
    });
  }

};

__decorate([property()], ThreeTexturesElement.prototype, "start", void 0);

__decorate([property()], ThreeTexturesElement.prototype, "end", void 0);

__decorate([property()], ThreeTexturesElement.prototype, "ext", void 0);

ThreeTexturesElement = __decorate([customElement('three-textures')], ThreeTexturesElement);
export { ThreeTexturesElement }; //# sourceMappingURL=three-textures.js.map
